File: MiddleElementOfLinkedList/MiddleElementOfLinkedList.cpp
// MiddleElementOfLinkedList.cpp : This file contains the 'main' function. Program execution begins and ends there.
//

#include <iostream>

class ListNode
{
public:
    int data;
    ListNode* next;
    ListNode(int data)
        : data(data)
        , next(nullptr)
    {}
};

ListNode* GetListFromInput(ListNode** ppHead, int count)
{
    if (count == 0)
        return nullptr;

    std::cout << "Enter " << count << " numbers of the list: ";
    ListNode* pTail = *ppHead;
    for (int i = 0; i < count; ++i)
    {
        int data = 0;
        std::cin >> data;
        if (*ppHead == nullptr)
        {
            *ppHead = new ListNode(data);
            pTail = *ppHead;
        }
        else
        {
            pTail->next = new ListNode(data);
            pTail = pTail->next;
        }
    }

    return pTail;
}

void PrintList(ListNode* pHead)
{
    std::cout << pHead->data;
    pHead = pHead->next;

    while (pHead != nullptr)
    {
        std::cout << ", " << pHead->data;
        pHead = pHead->next;
    }

    std::cout << std::endl;
}

void FreeList(ListNode* pNode)
{
    if (pNode == nullptr)
        return;

    FreeList(pNode->next);
    delete pNode;
}

int GetElementCount(ListNode* pNode)
{
    int count = 0;
    while (pNode)
    {
        ++count;
        pNode = pNode->next;
    }

    return count;
}

ListNode* GetNodeByIndex(ListNode* pHead, int index)
{
    if (index == -1)
        return nullptr;

    for (int i = 0; i < index; ++i)
    {
        pHead = pHead->next;
    }

    return pHead;
}

int GetMiddleElementOfLinkedList(ListNode* pHead)
{
    int count = GetElementCount(pHead);
    int index = (count - 1) / 2;
    return GetNodeByIndex(pHead, index)->data;
}

int main()
{
    while (true)
    {
        int count = 0;
        std::cout << "Number of nodes in list: ";
        std::cin >> count;
        if (count == 0)
            break;

        ListNode* pOriginalHead = nullptr;
        ListNode* pOriginalTail = GetListFromInput(&pOriginalHead, count);

        std::cout << "Original: ";
        PrintList(pOriginalHead);

        std::cout << "The middle element of the linked list is: " << GetMiddleElementOfLinkedList(pOriginalHead) << std::endl;
    }
}

File: Go/MiddleElementOfLinkedList.go
package main

import "fmt"

type ListNode struct {
	data int
	next *ListNode
}

func GetArrayFromInput(count int, prompt string) []int {
	if count == 0 {
		return []int{}
	}

	fmt.Print(prompt)

	v := make([]int, count)
	for i := 0; i < count; i++ {
		fmt.Scan(&v[i])
	}

	return v
}

func GetListFromArray(v []int) *ListNode {
	if len(v) == 0 {
		return nil
	}

	if len(v) == 1 {
		return &ListNode{v[0], nil}
	}

	pHead := &ListNode{v[0], nil}
	pCur := pHead
	for i := 1; i < len(v); i++ {
		pCur.next = &ListNode{v[i], nil}
		pCur = pCur.next
	}

	return pHead
}

func CountLinkedList(list *ListNode) int {
	count := 0
	for ; list != nil; list = list.next {
		count++
	}

	return count
}

func GetNodeFromIndex(list *ListNode, index int) *ListNode {
	for i := 0; i < index; i++ {
		list = list.next
	}

	return list
}

func GetMiddleElementOfLinkedList(list *ListNode) int {
	count := CountLinkedList(list)
	index := (count - 1) / 2

	return GetNodeFromIndex(list, index).data
}

func main() {
	for {
		fmt.Print("Number of elements: ")
		var count int
		fmt.Scan(&count)
		if count == 0 {
			break
		}

		vec := GetArrayFromInput(count, "Please enter the list: ")
		pHead := GetListFromArray(vec)

		fmt.Print("The middle element of the linked list is ")
		fmt.Println(GetMiddleElementOfLinkedList(pHead))
	}
}
